contributors/design-proposals/api-machinery/client-go-context.md

```diff
@@ -0,0 +1,141 @@
+	result = &v1.Pod{}
+	err = c.client.Get().
+		Namespace(c.ns).
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Context(ctx).
+		Do().
+		Into(result)
+	return
+}
+```
+
+This would result in an API that would be interacted with like so if the consumer wished to pass a context:
+```go  
+   ctx := req.Context()
+   pod, err := k8client.CoreV1().Pods(namespace).WithContext(ctx).Get(podName, ...)
+```
+
+
+HTTP Tracing and cancellation are also likely useful when using ```informers``` and ```listers```, as these also use the client interfaces backwards compatibility can be maintained by adding a new set of constructors for adding a
+specific context object:
+
+```go
+func NewConfigMapInformerWithContext(client kubernetes.Interface, namespace string, resyncPeriod time.Duration, indexers cache.Indexers, ctx context.Context) cache.SharedIndexInformer {}
+
+func (f *configMapInformer) InformerWithContext(ctx context.Context) cache.SharedIndexInformer {}
+
+func defaultConfigMapInformerWithContext(client kubernetes.Interface, resyncPeriod time.Duration, ctx context.Context) cache.SharedIndexInformer {
+	return NewConfigMapInformerWithContext(client, meta_v1.NamespaceAll, resyncPeriod, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},ctx)
+}
+
+func (f *configMapInformer) ListerWithContext(ctx context.Context) v1.ConfigMapLister {
+	return v1.NewConfigMapLister(f.InformerWithContext(ctx).GetIndexer())
+}
+```
+
+**Concerns / Notes**
+- The idea behind using a context with a request is it is only meant to live as long as the initial request that created it. If we expose an API like the one outlined, the consumer would need to know that they needed to call
+```WithContext(ctx context.Context)``` with each request context.
+
+- The fakes would need to be updated and new tests added to cover the new APIs.
+
+- The number of entry points and maintenance points is increased. 
+
+### Breaking change
+
+The second option is to pass the context with each action on a resource. This may be more idiomatic and explicit: 
+
+```go 
+func (c *pods) Get(ctx context.Context, name string, options meta_v1.GetOptions) (result *v1.Pod, err error) {
+	result = &v1.Pod{}
+	err = c.client.Get().
+		Namespace(c.ns).
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Context(c.ctx).
+		Do().
+		Into(result)
+	return
+}
+```
+
+These change would impact on every API. Although the changes required would be many, the simplest approach to maintaining current behaviour would be to pass in nil or ```context.ToDo()```.
+
+For the ```listers``` and ```informers``` the change outlined above (adding new constructors) would still allow a context to be passed through to client functions doing the work.
+
+**Concerns**
+
+This is very clearly a breaking change that would likely require extensive changes to both tests and implementation code throughout the impacted code bases.
```

Comment:
I don't care much about everything in `k8s.io/kubernetes`. It's a one-time cleanup to adapt using `context.ToDo()`. 

Passing a context through where it makes sense would be a bigger change, but that's not dependent on which of your solution we choose.

It's also a breaking change for our published client-go. We actually had similar changes in 1.6 with GetOptions and ListOptions. It might be tedious for our 3rdparty users to fix their code. But at least the breakage and fix is obvious and easy to repair.

Reply:
Agree that is it is relatively easy to fix. Also clearly announces context support in the client. With the none breaking change, a consumer would need to discover that context support was now available.

Reply:
On the other hand, this approach would guarantee that we update all call sites.  The other we could silently miss them.

Reply:
silently missing call site updates seems bad

Reply:
I'd be more concerned with third-parties scrambling to fix their build and misusing context without. If we silent'y failed back to `context.ToDo()` people could add context on their own initiative.